hyperspy/io.py

```diff
@@ -848,3 +852,51 @@ def save(filename, signal, overwrite=None, **kwds):
                 signal.tmp_parameters.set_item('folder', file.parent)
                 signal.tmp_parameters.set_item('filename', file.stem)
                 signal.tmp_parameters.set_item('extension', extension)
+
+
+def add_file_load_metadata(signal_dict, reader):
+    mdata_dict = {
+        'operation': 'load',
+        'io_plugin': reader.__loader__.name,
+        'hyperspy_version': hs_version,
+        'timestamp': datetime.now().astimezone().isoformat()
+    }
+    if 'FileIO' not in signal_dict['metadata']['General']:
+        signal_dict['metadata']['General']['FileIO'] = {
+            '0': mdata_dict
+        }
+    else:
+        largest_index = max([
+            int(i.replace('Number_', '')) for i in
+            signal_dict['metadata']['General']['FileIO'].keys()])
+        signal_dict['metadata'][
+            'General']['FileIO'][str(largest_index + 1)] = mdata_dict
+
+    return signal_dict
+
+
+def add_file_save_metadata(signal, writer):
+    mdata_dict = {
+        'operation': 'save',
+        'io_plugin': writer.__loader__.name,
+        'hyperspy_version': hs_version,
+        'timestamp': datetime.now().astimezone().isoformat()
+    }
+    if 'General' not in signal.metadata or isinstance(signal.metadata, dict):
+        # if we don't even have basic metadata as a DictionaryTreeBrowser,
+        # just return as-is (shouldn't happen, but this came up in tests of
+        # test_nexus_hdf.test_save_metadata_as_dict())
+        return signal
+    if 'FileIO' not in signal.metadata.General:
+        signal.metadata.General.add_dictionary = {
+            '0': mdata_dict
+        }
+    else:
+        largest_index = \
```

Comment:
An alternative would be to create it, if this is not there and the `if`/`else` block is not necessary, see for example:
```python

import hyperspy.api as hs

s = hs.signals.Signal1D([0, 1, 2])

index = 0
mdict = {'a': 'A', 'b': 'B'}

s.metadata.set_item(f'General.FileIO.{index}', mdict)

del s.metadata.General
s.metadata.set_item(f'General.FileIO.{index}', mdict)


```

Reply:
in this implementation, we'd still need an `if/else` to control what "number" should be used ("0", "1", "2", etc.), so I don't think it would simplify much, but I can change it if you'd like

Reply:
Once you know the index to be used, you can use `s.metadata.set_item(f'General.FileIO.{index}', mdict)` regardless if `FileIO` already exist or not.

For example:
```python
import hyperspy.api as hs

s = hs.signals.Signal1D([0, 1, 2])
mdict = {'a': 'A', 'b': 'B'}

# do it 3 times 
for _ in range(2):
    largest_index = max([int(i.replace('Number_', '')) + 1 for i in
                     s.metadata.get_item('General.FileIO', {}).keys()] + [0, ])
    print('largest_index', largest_index)
    
    s.metadata.set_item(f'General.FileIO.{largest_index}', mdict)
```

Reply:
Clever ;)

That's what I get for commenting without looking at the actual code...